messages/src/mgmt/mod.rs

```diff
@@ -0,0 +1,222 @@
+/// including identifying the management specification, and which pages if any
+/// are supported for a module. The intention is for the SP to only interpret
+/// sections of the map insofar as required for temperature and power
+/// monitoring.
+///
+/// Note that modules generally don't _fail_ a memory access to an unsupported
+/// page. Instead, modules generally just revert the page- or bank-select bytes
+/// to those known to be valid, usually zero. That means the client _must_ keep
+/// track of which bank and/or page is being accessed, and validate that those
+/// are actually supported by the module prior to operating on them.
+#[derive(Clone, Copy, Debug, Deserialize, PartialEq, Serialize, SerializedSize)]
+pub enum UpperPage {
+    Sff8636(sff8636::Page),
+    Cmis(cmis::Page),
+}
+
+impl UpperPage {
+    /// Return the actual page number of the upper page being mapped in.
+    pub fn page(&self) -> u8 {
+        match self {
+            UpperPage::Sff8636(inner) => inner.page(),
+            UpperPage::Cmis(inner) => inner.page(),
+        }
+    }
+
+    /// Return the bank of the upper page being accessed.
+    ///
+    /// Note that this returns `None` for pages that don't have a bank, e.g.,
+    /// those for modules conforming to SFF-8636.
+    ///
+    /// For those which _may_ have a bank, this _may_ return `Some(_)`. Note
+    /// that not all CMIS-defined pages allow a bank, so it may still return
+    /// `None` in that case.
+    pub fn bank(&self) -> Option<u8> {
+        match self {
+            UpperPage::Sff8636(_) => None,
+            UpperPage::Cmis(inner) => inner.bank(),
+        }
+    }
+}
+
+impl From<sff8636::Page> for UpperPage {
+    fn from(p: sff8636::Page) -> Self {
+        Self::Sff8636(p)
+    }
+}
+
+impl From<cmis::Page> for UpperPage {
+    fn from(p: cmis::Page) -> Self {
+        Self::Cmis(p)
+    }
+}
+
+/// A description of a region of a transceiver's memory map.
+///
+/// See [`UpperPage`] for a detailed description of how a module's upper page is
+/// mapped in for an operation. This type also performs validation against the
+/// offset / length, to ensure they're a valid access for the 256-byte memory
+/// map of all transceiver modules we support.
+#[derive(Clone, Copy, Debug, Deserialize, PartialEq, Serialize, SerializedSize)]
+pub struct MemoryRegion {
+    upper_page: UpperPage,
+    offset: u8,
+    len: u8,
+}
+
+impl MemoryRegion {
+    /// Construct a new memory region.
+    ///
+    /// Note that you must specify an `upper_page`, even if you are only
+    /// accessing the fixed, lower page of memory. In that case, you can use
+    /// the default for the specification of the accessed module, which uses
+    /// page and / or bank zero.
+    pub fn new(upper_page: UpperPage, offset: u8, len: u8) -> Result<Self, Error> {
+        // The last accessed byte must be within the 256-byte memory map.
+        if offset.checked_add(len).is_none() {
```

Comment:
There may be an off-by-one issue here:

If `offset = 250` and `len = 6`, we _should_ be able to read bytes 250-255 (inclusive), but `checked_add` will give `None` (and hence this will return an error).

Reply:
Ah, nice catch, thanks. I was trying to avoid a bunch of conversions or casts, but I think that might actually be the clearest mechanism. I'll sort it, and add a regression. Thank you.

Reply:
Resolved in https://github.com/oxidecomputer/transceiver-control/pull/5/commits/9aa5afd9d3e88eeab9164143d05a957d56588650